File: app/src/main/java/rogue_opcode/containers/Container.java
// Container.java
// Dynamical array-based container abstract base class
//
// Copyright © 2009-2017 Brigham Toskin
// This software is part of the Rogue-Opcode game framework. It is distributable
// under the terms of a modified MIT License. You should have received a copy of
// the license in the file LICENSE.md. If not, see:
// <http://code.google.com/p/rogue-op/wiki/LICENSE>
//
// Formatting:
//	80 cols ; tabwidth 4
////////////////////////////////////////////////////////////////////////////////


package rogue_opcode.containers;

import java.io.Serializable;
import java.util.Iterator;
import java.util.NoSuchElementException;


/** Simple {@code Exception} class that builds a friendly error message.
 * @see java.lang.Exception
 */
@SuppressWarnings("serial")
class ContainerError extends RuntimeException
{
	/** Constructs an {@code Exception} with a friendly error message, which
	 * includes the name of the calling method two frames back. This is because
	 * we don't usually throw one of these ourselves, but pawn it off to a
	 * helper method.
	 * @param error_msg brief description of error condition.
	 * @param context used to get stack trace.
	 */
	ContainerError(String error_msg, Throwable context)
	{
		super("Error in " + context.getStackTrace()[1].toString() + " - " +
				error_msg, context);
	}

	/**
	 * Constructs an {@code Exception} with a friendly error message, which
	 * includes the name of the calling method two frames back. This is because
	 * we don't usually throw one of these ourselves, but pawn it off to a
	 * helper method.
	 *
	 * @param error_msg brief description of error condition.
	 */
	ContainerError(String error_msg)
	{
		this(error_msg, null);
	}
}


/**
 * {@code Container} is an abstract base class for array-based data structures.
 * Efficiency in storage, access, and memory management are the primary design
 * goals. {@code Container}'s <i>raison d'être</i> is to provide a fast access
 * to self-sizing collections of data elements suitable for real-time game
 * applications.
 * <br /><br />
 * With this in mind, all data members have been
 * declared public and accessors should be utilized judiciously at runtime. When
 * accessing the {@code data} array directly, you should always keep in mind
 * that the length of the array is not necessarily the same as the logical size
 * of the container. Thus, you should check both values when calculating range
 * iterations.
 * <br /><br />
 * Note that {@code Container} is an abstract class; some of the
 * common and default interfaces have been implemented for you. To use
 * {@code Container}, extend it with your derived class, and implement the
 * missing
 * methods and any others that require customized behavior, <i>e.g.</i> for a
 * circular queue.
 *
 * @param <E> Generic storage type parameter.
 * @author Brigham Toskin
 */
public abstract class Container<E> implements Serializable, Iterable<E>
{
	private static final long serialVersionUID = -3657997720954461476L;

	public int size;
	public E[] data;

	public Container(int pCapacity)// throws Exception
	{
		//		Clear();
		Reserve(pCapacity);
	}

	// memory management ///////////////////////////////////////////////////////

	public E[] Data()
	{
		return data;
	}

	@SuppressWarnings("unchecked")
	public void Reserve(int pCount)
	{
		E[] tData = data; // grab a local ref

		try
		{
			if(tData != null) // already have a data buffer
			{
				if(pCount <= tData.length)
					return; // bail if capacity grater than the requested size
				// copy into larger buffer
				E[] tNewData = (E[])new Object[pCount];
				for(int i = 0; i < tData.length; i++)
					tNewData[i] = tData[i];
				tData = tNewData;
			}
			else
				tData = (E[])new Object[pCount]; // allocate requested buffer
		}
		catch(OutOfMemoryError e)
		{
			throw new ContainerError("Allocation failure", e);
		}

		//		validate_alloc(tData);
		data = tData;
	}


	/** Removes an arbitrary element from the middle of the container.<br />
	 * <b>Linear copy is inefficient; you probably should not be doing this.</b>
	 * @param pIndex index of element to remove.
	 */
	public void Remove(int pIndex)
	{
		validate_index(pIndex);
		E[] tData = data; // grab local ref
		tData[pIndex] = null; // "delete" the item
		for(; pIndex < size - 1; pIndex++) // copy  down if necessary
			tData[pIndex] = tData[pIndex + 1];
		size--;
	}
	public void Clear()
	{
		int tCapacity = data.length;
		data = null;
		size = 0;
		try
		{
			Reserve(tCapacity);
		}
		catch(Exception e)
		{
		}
	}
	public boolean Empty() { return size == 0; }

	// element access aliases //////////////////////////////////////////////////

	public abstract E First();;

	public abstract E Last();;

	public abstract E At(int pIndex);;

	protected class ContainerIterator implements Iterator<E>
	{
		int mIndex = 0;
		boolean mDirty = true;

		@Override
		public boolean hasNext()
		{
			return mIndex < size;
		}

		@Override
		public E next()
		{
			mDirty = false;
			if(!hasNext())
				throw new NoSuchElementException("End of container.");
			return data[mIndex++];
		}

		@Override
		public void remove()
		{
			if(mDirty)
				throw new IllegalStateException("remove() called before next()");
			Remove(--mIndex);
		}
	}

	@Override
	public Iterator<E> iterator()
	{
		return new ContainerIterator();
	}

	// error checking and memory helpers ///////////////////////////////////////

	protected static void validate_pointer(Object pRef)
	{
		if(pRef == null)
			throw new ContainerError("Dereferencing null pointer.",
					new Exception());
	}

	protected static void validate_alloc(Object pRef)
	{
		if(pRef == null)
			throw new ContainerError("Allocation failed.", new Exception());
	}

	protected void validate_nonempty()
	{
		if(size < 1)
			throw new ContainerError("Accessing empty container.",
					new Exception());
	}

	protected void validate_index(int pIndex)
	{
		if(pIndex >= size)
			throw new ContainerError("Index out of bounds.", new Exception());
	}

	protected void resize_inc()
	{
		int tLen = data.length;
		if(size == tLen)
			Reserve(tLen << 1);
		size++;
	}

	protected void resize_add(int count)
	{
		if(size + count > data.length)
		{
			int cap = data.length;
			do cap <<= 1; while(size + count > cap); // find a power of 2 size
			Reserve(cap);
		}
	}
}

File: lib/src/main/java/rogue_opcode/containers/Container.kt
// Container.kt
// Dynamical array-based container abstract base class
//
// Copyright © 2009-2017 Brigham Toskin
// This software is part of the Rogue-Opcode game framework. It is distributable
// under the terms of a modified MIT License. You should have received a copy of
// the license in the file LICENSE.md. If not, see:
// <https://github.com/IonoclastBrigham/RogueOp/blob/master/LICENSE.md>
//
// Formatting:
//	80 cols ; tabwidth 4
////////////////////////////////////////////////////////////////////////////////


package rogue_opcode.containers

import java.io.Serializable
import java.util.*
import kotlin.Array


/**
 * *Simple `Exception` class that builds a friendly error message.*
 * @see java.lang.Exception
 */
internal class ContainerError
/**
 * *Constructs an `Exception` with a friendly error message.*
 *
 * Message includes the name of the calling method two frames back. This is because
 * we don't usually throw one of these ourselves, but pawn it off to a
 * helper method.
 *
 * @param msg brief description of error condition.
 * @param context used to get stack trace.
 */
@JvmOverloads
constructor(msg: String, context: Throwable)
	: RuntimeException("Error in " + context.stackTrace[1].toString() + " - " + msg) {
	init {
		stackTrace = context.stackTrace.run { sliceArray(1..lastIndex) }
	}
}


/**
 * `Container` is an abstract base class for array-based mData structures.
 * Efficiency in storage, access, and memory management are the primary design
 * goals. `Container`'s *raison d'être* is to provide a fast access
 * to self-sizing collections of mData elements suitable for real-time game
 * applications.
 * <br></br><br></br>
 * With this in mind, all mData members have been
 * declared public and accessors should be utilized judiciously at runtime. When
 * accessing the `mData` array directly, you should always keep in mind
 * that the length of the array is not necessarily the same as the logical mSize
 * of the container. Thus, you should check both values when calculating range
 * iterations.
 * <br></br><br></br>
 * Note that `Container` is an abstract class; some of the
 * common and default interfaces have been implemented for you. To use
 * `Container`, extend it with your derived class, and implement the
 * missing
 * methods and any others that require customized behavior, *e.g.* for a
 * circular queue.

 * @param <E> Generic storage type parameter.
 * @author Brigham Toskin
 */
@Suppress("UNCHECKED_CAST")
abstract class Container<E>(pCapacity: Int) : Serializable, Iterable<E> {
	companion object {
		private const val serialVersionUID = -3657997720954461476L

		inline fun nextPow2(pVal: Int) = (pVal - 1).let { tVal ->
			tVal.or(tVal shr 1)
				.or(tVal shr 2)
				.or(tVal shr 4)
				.or(tVal shr 8)
				.or(tVal shr 16)
		} + 1
	}


	protected var mData = arrayOfNulls<Any?>(pCapacity)
	val Data get() = mData as Array<out E?>

	protected var mSize = 0
	val Size get() = mSize

	val LastIndex get() = mSize - 1

	// memory management ///////////////////////////////////////////////////////

	open fun Reserve(pCount: Int) {
		if(pCount <= mSize) return

		mData = try {
			Array(pCount) { i -> if(i in mData.indices) mData[i] else null }
		} catch(e: OutOfMemoryError) {
			throw ContainerError("Allocation failure", e)
		}
	}

	/** *Removes an arbitrary element from the middle of the container.*
	 *
	 * **Linear copy is inefficient; you probably should not be doing this.**
	 * @param pIndex index of element to remove.
	 */
	open fun Remove(pIndex: Int) {
		validate_index(pIndex)
		mData[pIndex] = null // "delete" the item
		val tLastSafeIndex = LastIndex - 1
		for(i in pIndex..tLastSafeIndex) {
			// copy  down if necessary
			mData[i] = mData[i + 1]
		}
		mData[LastIndex] = null
		mSize--
	}

	open fun Clear() {
		for(i in 0..LastIndex) mData[i] = null
		mSize = 0
	}

	fun Empty(): Boolean {
		return mSize == 0
	}

	// element access aliases //////////////////////////////////////////////////

	abstract fun First(): E

	abstract fun Last(): E

	abstract fun At(pIndex: Int): E

	protected open inner class ContainerIterator : MutableIterator<E> {
		internal var mIndex = 0
		internal var mDirty = true

		override fun hasNext() = mIndex < mSize

		override fun next(): E {
			mDirty = false
			if(!hasNext()) throw NoSuchElementException("End of container.")
			return mData[mIndex++] as E
		}

		override fun remove() {
			if(mDirty) throw IllegalStateException("remove() called before next()")
			Remove(--mIndex)
		}
	}

	override fun iterator(): MutableIterator<E> {
		return ContainerIterator()
	}

	protected fun validate_nonempty() {
		if(mSize < 1) {
			val ex = Exception()
			throw ContainerError("Accessing empty container.", ex)
		}
	}

	protected fun validate_index(pIndex: Int) {
		if(pIndex >= mSize) {
			val ex = Exception()
			throw ContainerError("Index out of bounds.", ex)
		}
	}

	protected fun resize_inc() {
		if(mData.size == mSize) Reserve(mSize * 2)
		mSize++
	}

	protected fun resize_add(pCount: Int) {
		val tRequired = mSize + pCount
		var tCapacity = mData.size
		if(tRequired > tCapacity) {
			Reserve(nextPow2(tRequired))
		}
	}
}
